### DISCUSSION TITLE
how to query any synset relation type, like 'agent', 'causes', etc ?

### DISCUSSION THREAD

**Discussion by costaverde** [goodmami/wn#134] [Q&A] [answered]
Hi,

Many synset relation types are defined as wn.constants.SYNSET_RELATIONS in 
https://wn.readthedocs.io/en/latest/api/wn.constants.html, but it seems we can only get the most common types like hypernyms, hyponyms, etc.

Is there a way to query any of them, maybe via Sqlite and how ?

Thanks

**Comment by vsraptor** [comment]
https://github.com/goodmami/wn/blob/main/wn/_queries.py

i tracked this as a root of .similarity calculations:

https://github.com/goodmami/wn/blob/f587a33ce5cbef165a379e46f7df1c5bf054b034/wn/_core.py#L459  

which is called by : 

        paths = list(self.relation_paths('hypernym', 'instance_hypernym'))

i'm also looking of what are other ways of relations

**Comment by costaverde** [comment]
Yes, in https://github.com/goodmami/wn/blob/main/wn/_queries.py
there is 
def get_synset_relations(
    source_rowids: Collection[int],
    relation_types: Collection[str],
    lexicon_rowids: Sequence[int],
) -> Iterator[_Synset_Relation]:
doing the job. 
Looking for how to use it..

**Comment by vsraptor** [comment]
you can also use SQLite browser with ~/.wn_data/wn.db

**Comment by goodmami** [comment, marked as answer]
The `wn._queries` module is not part of the public API, nor is any other module prefixed with `_`. This means that the functions are not documented and they may (and do, occasionally) change from one version to the next without a CHANGELOG entry or change to the major version number. I do not recommend using those.

There is the [Synset.get_related()](https://wn.readthedocs.io/en/latest/api/wn.html#wn.Synset.get_related) method for finding relations with any available relation name. Similarly for senses is [Sense.get_related()](https://wn.readthedocs.io/en/latest/api/wn.html#wn.Sense.get_related), and for sense-to-synset relations there is [Sense.get_related_synsets()](https://wn.readthedocs.io/en/latest/api/wn.html#wn.Sense.get_related_synsets). Since the last release, both `Synset` and `Sense` also have a `relations()` method which does the same thing but returns a dictionary mapping the relation names to the lists of relations, but unfortunately I forgot to document these. They will be in the next release. The [Synset.relation_paths()](https://wn.readthedocs.io/en/latest/api/wn.html#wn.Synset.relation_paths) and [Sense.relation_paths()](https://wn.readthedocs.io/en/latest/api/wn.html#wn.Sense.relation_paths) methods return a list of lists of synsets/senses connected by one or more relations.

Methods like [Synset.hypernyms()](https://wn.readthedocs.io/en/latest/api/wn.html#wn.Synset.hypernyms) are just for convenience and they capture multiple individual relations. For instance, `hypernyms()` includes both the `hypernym` and `instance_hypernym` relations:

https://github.com/goodmami/wn/blob/f587a33ce5cbef165a379e46f7df1c5bf054b034/wn/_core.py#L896-L906

Also, as a general note, it is not the case that all relations in `wn.constants.SYNSET_RELATIONS` are in use by any one wordnet. A subset of those are used by the English wordnets, another subset by the Polish wordnet, etc.

**Comment by vsraptor** [comment]
thanks thats very helpful ..i'm trying to figure the relationship structure so I can figure different ways of word-similarity..

Monkey patched __repr__ so I can follow it easier : 

```
def synset_repr(self):
  return re.sub(r"[\[\]']", '', str(self.lemmas()) ).replace(', ',':')

wn.Synset.__repr__ = synset_repr
```

**Comment by costaverde** [comment]
Thanks very much @goodmami 

I missed the  https://wn.readthedocs.io/en/latest/api/wn.html#wn.Synset.get_related in  the doc.

**Comment by costaverde** [comment]
[](url)
